File: sdk/py/hop_top_kit/toolspec.py
"""Structured knowledge about CLI tools — Python port of Go toolspec."""

from __future__ import annotations

from dataclasses import dataclass, field

import yaml

# --- Types ---


@dataclass
class Contract:
    idempotent: bool = False
    side_effects: list[str] = field(default_factory=list)
    retryable: bool = False
    pre_conditions: list[str] = field(default_factory=list)


@dataclass
class Safety:
    level: str = "safe"
    requires_confirmation: bool = False
    permissions: list[str] = field(default_factory=list)


@dataclass
class OutputSchema:
    format: str = ""
    fields: list[str] = field(default_factory=list)
    example: str = ""


@dataclass
class StateIntrospection:
    config_commands: list[str] = field(default_factory=list)
    env_vars: list[str] = field(default_factory=list)
    auth_commands: list[str] = field(default_factory=list)


@dataclass
class Provenance:
    source: str = ""
    retrieved_at: str = ""
    confidence: float = 0.0


@dataclass
class Intent:
    domain: str = ""
    category: str = ""
    tags: list[str] = field(default_factory=list)


@dataclass
class Flag:
    name: str = ""
    short: str = ""
    type: str = ""
    description: str = ""
    deprecated: bool = False
    replaced_by: str = ""


@dataclass
class Command:
    name: str = ""
    aliases: list[str] = field(default_factory=list)
    flags: list[Flag] | None = None
    children: list[Command] | None = None
    contract: Contract | None = None
    safety: Safety | None = None
    preview_modes: list[str] | None = None
    output_schema: OutputSchema | None = None
    deprecated: bool = False
    deprecated_since: str = ""
    replaced_by: str = ""
    intent: Intent | None = None
    suggested_next: list[str] | None = None


@dataclass
class ErrorPattern:
    pattern: str = ""
    fix: str = ""
    source: str = ""
    cause: str = ""
    fixes: list[str] = field(default_factory=list)
    confidence: float = 0.0
    provenance: Provenance | None = None


@dataclass
class Workflow:
    name: str = ""
    steps: list[str] = field(default_factory=list)
    after: dict[str, list[str]] | None = None
    provenance: Provenance | None = None


@dataclass
class ToolSpec:
    name: str = ""
    schema_version: str = ""
    commands: list[Command] = field(default_factory=list)
    flags: list[Flag] | None = None
    error_patterns: list[ErrorPattern] | None = None
    workflows: list[Workflow] | None = None
    state_introspection: StateIntrospection | None = None


# --- YAML mapping helpers ---


def _map_flag(raw: dict) -> Flag:
    return Flag(
        name=raw.get("name", ""),
        short=raw.get("short", ""),
        type=raw.get("type", ""),
        description=raw.get("description", ""),
        deprecated=raw.get("deprecated", False),
        replaced_by=raw.get("replaced_by", ""),
    )


def _map_provenance(raw: dict | None) -> Provenance | None:
    if not raw:
        return None
    return Provenance(
        source=raw.get("source", ""),
        retrieved_at=raw.get("retrieved_at", ""),
        confidence=raw.get("confidence", 0.0),
    )


def _map_contract(raw: dict | None) -> Contract | None:
    if not raw:
        return None
    return Contract(
        idempotent=raw.get("idempotent", False),
        side_effects=raw.get("side_effects", []),
        retryable=raw.get("retryable", False),
        pre_conditions=raw.get("pre_conditions", []),
    )


def _map_safety(raw: dict | None) -> Safety | None:
    if not raw:
        return None
    return Safety(
        level=raw.get("level", "safe"),
        requires_confirmation=raw.get("requires_confirmation", False),
        permissions=raw.get("permissions", []),
    )


def _map_output_schema(raw: dict | None) -> OutputSchema | None:
    if not raw:
        return None
    return OutputSchema(
        format=raw.get("format", ""),
        fields=raw.get("fields", []),
        example=raw.get("example", ""),
    )


def _map_intent(raw: dict | None) -> Intent | None:
    if not raw:
        return None
    return Intent(
        domain=raw.get("domain", ""),
        category=raw.get("category", ""),
        tags=raw.get("tags", []),
    )


def _map_command(raw: dict) -> Command:
    flags_raw = raw.get("flags")
    children_raw = raw.get("children")
    return Command(
        name=raw.get("name", ""),
        aliases=raw.get("aliases", []),
        flags=([_map_flag(f) for f in flags_raw] if flags_raw else None),
        children=([_map_command(c) for c in children_raw] if children_raw else None),
        contract=_map_contract(raw.get("contract")),
        safety=_map_safety(raw.get("safety")),
        preview_modes=raw.get("preview_modes"),
        output_schema=_map_output_schema(raw.get("output_schema")),
        deprecated=raw.get("deprecated", False),
        deprecated_since=raw.get("deprecated_since", ""),
        replaced_by=raw.get("replaced_by", ""),
        intent=_map_intent(raw.get("intent")),
        suggested_next=raw.get("suggested_next"),
    )


def _map_state_introspection(
    raw: dict | None,
) -> StateIntrospection | None:
    if not raw:
        return None
    return StateIntrospection(
        config_commands=raw.get("config_commands", []),
        env_vars=raw.get("env_vars", []),
        auth_commands=raw.get("auth_commands", []),
    )


def _map_error_pattern(raw: dict) -> ErrorPattern:
    return ErrorPattern(
        pattern=raw.get("pattern", ""),
        fix=raw.get("fix", ""),
        source=raw.get("source", ""),
        cause=raw.get("cause", ""),
        fixes=raw.get("fixes", []),
        confidence=raw.get("confidence", 0.0),
        provenance=_map_provenance(raw.get("provenance")),
    )


def _map_workflow(raw: dict) -> Workflow:
    return Workflow(
        name=raw.get("name", ""),
        steps=raw.get("steps", []),
        after=raw.get("after"),
        provenance=_map_provenance(raw.get("provenance")),
    )


# --- Public API ---


def load_toolspec(path: str) -> ToolSpec:
    """Parse a YAML toolspec file into a ToolSpec object."""
    with open(path) as f:
        raw = yaml.safe_load(f)

    commands_raw = raw.get("commands", [])
    error_patterns_raw = raw.get("error_patterns")
    workflows_raw = raw.get("workflows")
    flags_raw = raw.get("flags")

    return ToolSpec(
        name=raw.get("name", ""),
        schema_version=str(raw.get("schema_version", "")),
        commands=[_map_command(c) for c in commands_raw],
        flags=([_map_flag(f) for f in flags_raw] if flags_raw else None),
        error_patterns=(
            [_map_error_pattern(e) for e in error_patterns_raw] if error_patterns_raw else None
        ),
        workflows=([_map_workflow(w) for w in workflows_raw] if workflows_raw else None),
        state_introspection=_map_state_introspection(raw.get("state_introspection")),
    )


_VALID_SAFETY_LEVELS = {"safe", "caution", "dangerous"}


def validate_toolspec(spec: ToolSpec) -> list[str]:
    """Validate a ToolSpec, returning a list of errors."""
    errors: list[str] = []

    if not spec.name:
        errors.append("name is required")
    if not spec.schema_version:
        errors.append("schema_version is required")
    if not spec.commands:
        errors.append("at least one command is required")

    def _validate_cmd(cmd: Command, path: str) -> None:
        if not cmd.name:
            errors.append(f"{path}: command name is required")
        if cmd.safety and cmd.safety.level not in _VALID_SAFETY_LEVELS:
            errors.append(
                f"{path}: invalid safety level "
                f'"{cmd.safety.level}" '
                f"(must be safe|caution|dangerous)"
            )
        if cmd.children:
            for child in cmd.children:
                _validate_cmd(
                    child,
                    f"{path}/{child.name or '<unnamed>'}",
                )

    for cmd in spec.commands or []:
        _validate_cmd(cmd, cmd.name or "<unnamed>")

    return errors


def find_command(spec: ToolSpec, name: str) -> Command | None:
    """BFS search for a command by name (mirrors Go)."""
    queue = list(spec.commands)
    while queue:
        c = queue.pop(0)
        if c.name == name:
            return c
        if c.children:
            queue.extend(c.children)
    return None

File: sdk/ts/src/toolspec.ts
import * as fs from "fs";
import * as yaml from "js-yaml";

// --- Types ---

export type SafetyLevel = "safe" | "caution" | "dangerous";

export interface Contract {
  idempotent: boolean;
  sideEffects?: string[];
  retryable: boolean;
  preConditions?: string[];
}

export interface Safety {
  level: SafetyLevel;
  requiresConfirmation?: boolean;
  permissions?: string[];
}

export interface OutputSchema {
  format?: string;
  fields?: string[];
  example?: string;
}

export interface StateIntrospection {
  configCommands?: string[];
  envVars?: string[];
  authCommands?: string[];
}

export interface Provenance {
  source?: string;
  retrievedAt?: string;
  confidence?: number;
}

export interface Intent {
  domain?: string;
  category?: string;
  tags?: string[];
}

export interface Flag {
  name: string;
  short?: string;
  type?: string;
  description?: string;
  deprecated?: boolean;
  replacedBy?: string;
}

export interface Command {
  name: string;
  aliases?: string[];
  flags?: Flag[];
  children?: Command[];
  contract?: Contract;
  safety?: Safety;
  previewModes?: string[];
  outputSchema?: OutputSchema;
  deprecated?: boolean;
  deprecatedSince?: string;
  replacedBy?: string;
  intent?: Intent;
  suggestedNext?: string[];
}

export interface ErrorPattern {
  pattern: string;
  fix: string;
  source?: string;
  cause?: string;
  fixes?: string[];
  confidence?: number;
  provenance?: Provenance;
}

export interface Workflow {
  name: string;
  steps: string[];
  after?: Record<string, string[]>;
  provenance?: Provenance;
}

export interface ToolSpec {
  name: string;
  schemaVersion: string;
  commands: Command[];
  flags?: Flag[];
  errorPatterns?: ErrorPattern[];
  workflows?: Workflow[];
  stateIntrospection?: StateIntrospection;
}

// --- YAML mapping helpers ---

// snake_case YAML -> camelCase TS
function mapFlag(raw: any): Flag {
  return {
    name: raw.name ?? "",
    short: raw.short,
    type: raw.type,
    description: raw.description,
    deprecated: raw.deprecated,
    replacedBy: raw.replaced_by,
  };
}

function mapProvenance(raw: any): Provenance | undefined {
  if (!raw) return undefined;
  return {
    source: raw.source,
    retrievedAt: raw.retrieved_at,
    confidence: raw.confidence,
  };
}

function mapContract(raw: any): Contract | undefined {
  if (!raw) return undefined;
  return {
    idempotent: raw.idempotent ?? false,
    sideEffects: raw.side_effects,
    retryable: raw.retryable ?? false,
    preConditions: raw.pre_conditions,
  };
}

function mapSafety(raw: any): Safety | undefined {
  if (!raw) return undefined;
  return {
    level: raw.level,
    requiresConfirmation: raw.requires_confirmation,
    permissions: raw.permissions,
  };
}

function mapOutputSchema(raw: any): OutputSchema | undefined {
  if (!raw) return undefined;
  return {
    format: raw.format,
    fields: raw.fields,
    example: raw.example,
  };
}

function mapIntent(raw: any): Intent | undefined {
  if (!raw) return undefined;
  return {
    domain: raw.domain,
    category: raw.category,
    tags: raw.tags,
  };
}

function mapCommand(raw: any): Command {
  return {
    name: raw.name ?? "",
    aliases: raw.aliases,
    flags: raw.flags?.map(mapFlag),
    children: raw.children?.map(mapCommand),
    contract: mapContract(raw.contract),
    safety: mapSafety(raw.safety),
    previewModes: raw.preview_modes,
    outputSchema: mapOutputSchema(raw.output_schema),
    deprecated: raw.deprecated,
    deprecatedSince: raw.deprecated_since,
    replacedBy: raw.replaced_by,
    intent: mapIntent(raw.intent),
    suggestedNext: raw.suggested_next,
  };
}

function mapStateIntrospection(
  raw: any,
): StateIntrospection | undefined {
  if (!raw) return undefined;
  return {
    configCommands: raw.config_commands,
    envVars: raw.env_vars,
    authCommands: raw.auth_commands,
  };
}

function mapErrorPattern(raw: any): ErrorPattern {
  return {
    pattern: raw.pattern ?? "",
    fix: raw.fix ?? "",
    source: raw.source,
    cause: raw.cause,
    fixes: raw.fixes,
    confidence: raw.confidence,
    provenance: mapProvenance(raw.provenance),
  };
}

function mapWorkflow(raw: any): Workflow {
  return {
    name: raw.name ?? "",
    steps: raw.steps ?? [],
    after: raw.after,
    provenance: mapProvenance(raw.provenance),
  };
}

// --- Public API ---

/** Parse a YAML toolspec file into a ToolSpec object. */
export function loadToolSpec(yamlPath: string): ToolSpec {
  const content = fs.readFileSync(yamlPath, "utf-8");
  const raw = yaml.load(content) as any;

  return {
    name: raw.name ?? "",
    schemaVersion: raw.schema_version ?? "",
    commands: (raw.commands ?? []).map(mapCommand),
    flags: raw.flags?.map(mapFlag),
    errorPatterns: raw.error_patterns?.map(mapErrorPattern),
    workflows: raw.workflows?.map(mapWorkflow),
    stateIntrospection: mapStateIntrospection(
      raw.state_introspection,
    ),
  };
}

const VALID_SAFETY_LEVELS: Set<string> = new Set([
  "safe",
  "caution",
  "dangerous",
]);

/** Validate a ToolSpec, returning a list of error strings. */
export function validateToolSpec(spec: ToolSpec): string[] {
  const errors: string[] = [];

  if (!spec.name) errors.push("name is required");
  if (!spec.schemaVersion) errors.push("schemaVersion is required");
  if (!spec.commands || spec.commands.length === 0) {
    errors.push("at least one command is required");
  }

  const validateCmd = (cmd: Command, path: string) => {
    if (!cmd.name) {
      errors.push(`${path}: command name is required`);
    }
    if (cmd.safety && !VALID_SAFETY_LEVELS.has(cmd.safety.level)) {
      errors.push(
        `${path}: invalid safety level "${cmd.safety.level}"` +
          ` (must be safe|caution|dangerous)`,
      );
    }
    if (cmd.children) {
      for (const child of cmd.children) {
        validateCmd(child, `${path}/${child.name || "<unnamed>"}`);
      }
    }
  };

  for (const cmd of spec.commands ?? []) {
    validateCmd(cmd, cmd.name || "<unnamed>");
  }

  return errors;
}

/** BFS search for a command by name (mirrors Go FindCommand). */
export function findCommand(
  spec: ToolSpec,
  name: string,
): Command | undefined {
  const queue: Command[] = [...spec.commands];
  while (queue.length > 0) {
    const c = queue.shift()!;
    if (c.name === name) return c;
    if (c.children) queue.push(...c.children);
  }
  return undefined;
}
